File: BacktestResult.java
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;

public class BacktestResult {
    private String strategyName;
    private BigDecimal initialCapital;
    private BigDecimal finalCapital;
    private List<Trade> trades;
    private List<BigDecimal> equityCurve;
    private String startDate;
    private String endDate;

    public BacktestResult(String strategyName, BigDecimal initialCapital) {
        this.strategyName = strategyName;
        this.initialCapital = initialCapital;
        this.finalCapital = initialCapital;
        this.trades = new ArrayList<>();
        this.equityCurve = new ArrayList<>();
    }

    public void addTrade(Trade trade) {
        trades.add(trade);
    }

    public void setFinalCapital(BigDecimal finalCapital) {
        this.finalCapital = finalCapital;
    }

    public void setEquityCurve(List<BigDecimal> equityCurve) {
        this.equityCurve = equityCurve;
    }

    public void setStartDate(String startDate) { this.startDate = startDate; }
    public void setEndDate(String endDate) { this.endDate = endDate; }

    public BigDecimal getTotalReturn() {
        return finalCapital.subtract(initialCapital)
                .divide(initialCapital, 6, RoundingMode.HALF_UP)
                .multiply(BigDecimal.valueOf(100));
    }

    public BigDecimal getTotalProfitLoss() {
        return finalCapital.subtract(initialCapital);
    }

    public int getTotalTrades() {
        return (int) trades.stream().filter(t -> t.getStatus() == Trade.Status.CLOSED).count();
    }

    public int getWinningTrades() {
        return (int) trades.stream()
                .filter(t -> t.getStatus() == Trade.Status.CLOSED)
                .filter(Trade::isWin).count();
    }

    public int getLosingTrades() {
        return getTotalTrades() - getWinningTrades();
    }

    public BigDecimal getWinRate() {
        if (getTotalTrades() == 0) return BigDecimal.ZERO;
        return BigDecimal.valueOf(getWinningTrades())
                .divide(BigDecimal.valueOf(getTotalTrades()), 4, RoundingMode.HALF_UP)
                .multiply(BigDecimal.valueOf(100));
    }

    public BigDecimal getAverageWin() {
        List<BigDecimal> wins = new ArrayList<>();
        for (Trade t : trades) {
            if (t.getStatus() == Trade.Status.CLOSED && t.isWin()) {
                wins.add(t.getProfitLoss());
            }
        }
        if (wins.isEmpty()) return BigDecimal.ZERO;
        BigDecimal sum = BigDecimal.ZERO;
        for (BigDecimal w : wins) sum = sum.add(w);
        return sum.divide(BigDecimal.valueOf(wins.size()), 2, RoundingMode.HALF_UP);
    }

    public BigDecimal getAverageLoss() {
        List<BigDecimal> losses = new ArrayList<>();
        for (Trade t : trades) {
            if (t.getStatus() == Trade.Status.CLOSED && !t.isWin()) {
                losses.add(t.getProfitLoss());
            }
        }
        if (losses.isEmpty()) return BigDecimal.ZERO;
        BigDecimal sum = BigDecimal.ZERO;
        for (BigDecimal l : losses) sum = sum.add(l);
        return sum.divide(BigDecimal.valueOf(losses.size()), 2, RoundingMode.HALF_UP);
    }

    public BigDecimal getProfitFactor() {
        BigDecimal grossProfit = BigDecimal.ZERO;
        BigDecimal grossLoss = BigDecimal.ZERO;
        for (Trade t : trades) {
            if (t.getStatus() == Trade.Status.CLOSED) {
                if (t.isWin()) {
                    grossProfit = grossProfit.add(t.getProfitLoss());
                } else {
                    grossLoss = grossLoss.add(t.getProfitLoss().abs());
                }
            }
        }
        if (grossLoss.compareTo(BigDecimal.ZERO) == 0) {
            return grossProfit.compareTo(BigDecimal.ZERO) > 0 ? BigDecimal.valueOf(999.99) : BigDecimal.ZERO;
        }
        return grossProfit.divide(grossLoss, 2, RoundingMode.HALF_UP);
    }

    public BigDecimal getMaxDrawdown() {
        if (equityCurve.isEmpty()) return BigDecimal.ZERO;

        BigDecimal peak = equityCurve.get(0);
        BigDecimal maxDrawdown = BigDecimal.ZERO;

        for (BigDecimal equity : equityCurve) {
            if (equity.compareTo(peak) > 0) {
                peak = equity;
            }
            BigDecimal drawdown = peak.subtract(equity)
                    .divide(peak, 6, RoundingMode.HALF_UP)
                    .multiply(BigDecimal.valueOf(100));
            if (drawdown.compareTo(maxDrawdown) > 0) {
                maxDrawdown = drawdown;
            }
        }
        return maxDrawdown;
    }

    public BigDecimal getSharpeRatio(int tradingDaysPerYear) {
        if (equityCurve.size() < 2) return BigDecimal.ZERO;

        // Calculate daily returns
        List<Double> dailyReturns = new ArrayList<>();
        for (int i = 1; i < equityCurve.size(); i++) {
            double ret = equityCurve.get(i).subtract(equityCurve.get(i - 1))
                    .divide(equityCurve.get(i - 1), 10, RoundingMode.HALF_UP)
                    .doubleValue();
            dailyReturns.add(ret);
        }

        // Mean return
        double sumReturns = 0;
        for (double r : dailyReturns) sumReturns += r;
        double meanReturn = sumReturns / dailyReturns.size();

        // Standard deviation
        double sumSqDiff = 0;
        for (double r : dailyReturns) {
            sumSqDiff += Math.pow(r - meanReturn, 2);
        }
        double stdDev = Math.sqrt(sumSqDiff / dailyReturns.size());

        if (stdDev == 0) return BigDecimal.ZERO;

        // Annualized Sharpe (assuming risk-free rate ~ 0 for simplicity)
        double sharpe = (meanReturn / stdDev) * Math.sqrt(tradingDaysPerYear);
        return BigDecimal.valueOf(sharpe).setScale(2, RoundingMode.HALF_UP);
    }

    public BigDecimal getBuyAndHoldReturn(List<PriceData> data) {
        if (data.isEmpty()) return BigDecimal.ZERO;
        BigDecimal first = data.get(0).getClose();
        BigDecimal last = data.get(data.size() - 1).getClose();
        return last.subtract(first).divide(first, 6, RoundingMode.HALF_UP).multiply(BigDecimal.valueOf(100));
    }

    public void printReport(List<PriceData> data) {
        String separator = "=".repeat(70);
        String thinSep = "-".repeat(70);

        System.out.println();
        System.out.println(separator);
        System.out.println("  BACKTEST REPORT: " + strategyName);
        System.out.println(separator);
        System.out.println();

        System.out.println("  Period:              " + startDate + " to " + endDate);
        System.out.println("  Initial Capital:     $" + initialCapital.setScale(2, RoundingMode.HALF_UP));
        System.out.println("  Final Capital:       $" + finalCapital.setScale(2, RoundingMode.HALF_UP));
        System.out.println();

        System.out.println(thinSep);
        System.out.println("  PERFORMANCE METRICS");
        System.out.println(thinSep);
        System.out.printf("  Total Return:        %s%%%n", getTotalReturn().setScale(2, RoundingMode.HALF_UP));
        System.out.printf("  Buy & Hold Return:   %s%%%n", getBuyAndHoldReturn(data).setScale(2, RoundingMode.HALF_UP));
        System.out.printf("  Total P/L:           $%s%n", getTotalProfitLoss().setScale(2, RoundingMode.HALF_UP));
        System.out.printf("  Sharpe Ratio:        %s%n", getSharpeRatio(252));
        System.out.printf("  Max Drawdown:        %s%%%n", getMaxDrawdown().setScale(2, RoundingMode.HALF_UP));
        System.out.printf("  Profit Factor:       %s%n", getProfitFactor());
        System.out.println();

        System.out.println(thinSep);
        System.out.println("  TRADE STATISTICS");
        System.out.println(thinSep);
        System.out.printf("  Total Trades:        %d%n", getTotalTrades());
        System.out.printf("  Winning Trades:      %d%n", getWinningTrades());
        System.out.printf("  Losing Trades:       %d%n", getLosingTrades());
        System.out.printf("  Win Rate:            %s%%%n", getWinRate().setScale(2, RoundingMode.HALF_UP));
        System.out.printf("  Average Win:         $%s%n", getAverageWin());
        System.out.printf("  Average Loss:        $%s%n", getAverageLoss());
        System.out.println();

        if (!trades.isEmpty()) {
            System.out.println(thinSep);
            System.out.println("  TRADE LOG");
            System.out.println(thinSep);
            int count = 1;
            for (Trade t : trades) {
                if (t.getStatus() == Trade.Status.CLOSED) {
                    System.out.printf("  #%-4d %s%n", count++, t);
                }
            }
        }

        System.out.println();
        System.out.println(separator);
    }

    public String getStrategyName() { return strategyName; }
    public BigDecimal getInitialCapital() { return initialCapital; }
    public BigDecimal getFinalCapital() { return finalCapital; }
    public List<Trade> getTrades() { return trades; }
    public List<BigDecimal> getEquityCurve() { return equityCurve; }
}

File: ios/StockAnalyzer/StockAnalyzer/Models/BacktestResult.swift
import Foundation

struct BacktestResult: Identifiable {
    let id = UUID()
    let strategyName: String
    let initialCapital: Decimal
    let finalCapital: Decimal
    let trades: [Trade]
    let equityCurve: [(String, Decimal)]

    var totalReturn: Decimal {
        guard initialCapital != 0 else { return 0 }
        return ((finalCapital - initialCapital) / initialCapital) * 100
    }

    var totalTrades: Int {
        trades.filter { $0.status == .closed }.count
    }

    var winningTrades: Int {
        trades.filter { $0.status == .closed && ($0.profitLoss ?? 0) > 0 }.count
    }

    var losingTrades: Int {
        trades.filter { $0.status == .closed && ($0.profitLoss ?? 0) <= 0 }.count
    }

    var winRate: Decimal {
        guard totalTrades > 0 else { return 0 }
        return (Decimal(winningTrades) / Decimal(totalTrades)) * 100
    }

    var totalProfit: Decimal {
        trades.filter { $0.status == .closed && ($0.profitLoss ?? 0) > 0 }
            .compactMap { $0.profitLoss }
            .reduce(0, +)
    }

    var totalLoss: Decimal {
        trades.filter { $0.status == .closed && ($0.profitLoss ?? 0) <= 0 }
            .compactMap { $0.profitLoss }
            .reduce(0, +)
    }

    var profitFactor: Decimal {
        let loss = abs(NSDecimalNumber(decimal: totalLoss).decimalValue)
        guard loss != 0 else { return totalProfit > 0 ? 999 : 0 }
        return totalProfit / loss
    }

    var averageWin: Decimal {
        guard winningTrades > 0 else { return 0 }
        return totalProfit / Decimal(winningTrades)
    }

    var averageLoss: Decimal {
        guard losingTrades > 0 else { return 0 }
        return totalLoss / Decimal(losingTrades)
    }

    var maxDrawdown: Decimal {
        guard !equityCurve.isEmpty else { return 0 }
        var peak: Decimal = equityCurve[0].1
        var maxDD: Decimal = 0

        for (_, equity) in equityCurve {
            if equity > peak {
                peak = equity
            }
            let drawdown = peak > 0 ? ((peak - equity) / peak) * 100 : 0
            if drawdown > maxDD {
                maxDD = drawdown
            }
        }
        return maxDD
    }

    var sharpeRatio: Double {
        let returns = equityCurve.enumerated().compactMap { index, item -> Double? in
            guard index > 0 else { return nil }
            let prev = NSDecimalNumber(decimal: equityCurve[index - 1].1).doubleValue
            let curr = NSDecimalNumber(decimal: item.1).doubleValue
            guard prev > 0 else { return nil }
            return (curr - prev) / prev
        }

        guard returns.count > 1 else { return 0 }

        let mean = returns.reduce(0, +) / Double(returns.count)
        let variance = returns.map { ($0 - mean) * ($0 - mean) }.reduce(0, +) / Double(returns.count - 1)
        let stdDev = sqrt(variance)

        guard stdDev > 0 else { return 0 }
        let annualized = (mean / stdDev) * sqrt(252.0)
        return annualized
    }
}

struct OptionsBacktestResult: Identifiable {
    let id = UUID()
    let strategyName: String
    let optionType: OptionType
    let initialCapital: Decimal
    let finalCapital: Decimal
    let trades: [OptionsTrade]

    var totalReturn: Decimal {
        guard initialCapital != 0 else { return 0 }
        return ((finalCapital - initialCapital) / initialCapital) * 100
    }

    var totalTrades: Int {
        trades.filter { $0.status == .closed }.count
    }

    var winRate: Decimal {
        let closed = trades.filter { $0.status == .closed }
        guard !closed.isEmpty else { return 0 }
        let wins = closed.filter { ($0.profitLoss ?? 0) > 0 }.count
        return (Decimal(wins) / Decimal(closed.count)) * 100
    }

    var totalPremiumCollected: Decimal {
        trades.compactMap { $0.profitLoss }.filter { $0 > 0 }.reduce(0, +)
    }

    var totalPremiumPaid: Decimal {
        trades.compactMap { $0.profitLoss }.filter { $0 < 0 }.reduce(0, +)
    }

    var netPremium: Decimal {
        trades.compactMap { $0.profitLoss }.reduce(0, +)
    }

    var averageDaysHeld: Double {
        let held = trades.compactMap { $0.daysHeld }
        guard !held.isEmpty else { return 0 }
        return Double(held.reduce(0, +)) / Double(held.count)
    }
}
